pkg/apis/core/validation/validation_declarative_test.go

```diff
@@ -0,0 +1,134 @@
+	minReadySecondsPath := field.NewPath("spec").Child("minReadySeconds")
+	selectorPath := field.NewPath("spec").Child("selector")
+	containersPath := field.NewPath("spec").Child("template").Child("spec").Child("containers")
+
+	tests := []struct {
+		name                    string
+		imperativeErrors        field.ErrorList
+		declarativeErrors       field.ErrorList
+		expectMismatches        bool
+		expectDetailsContaining []string
+	}{
+		{
+			name: "Matching subset of imperative w/ covered and declarative errors- no mismatch",
+			imperativeErrors: field.ErrorList{
+				field.Invalid(replicasPath, nil, "regular error A"),
+				createCoveredError(minReadySecondsPath, -1, "covered error B").WithOrigin("minimum"),
+				createCoveredError(replicasPath, nil, "covered error C").WithOrigin("minimum"),
+				field.Invalid(selectorPath, nil, "regular error D"),
+			},
+			declarativeErrors: field.ErrorList{
+				field.Invalid(minReadySecondsPath, -1, "covered error B").WithOrigin("minimum"),
+				field.Invalid(replicasPath, nil, "covered error C").WithOrigin("minimum"),
+			},
+			expectMismatches:        false,
+			expectDetailsContaining: []string{},
+		},
+		{
+			name: "Non-matching subset of imperative w/ covered and declarative errors- missing declarative error",
+			imperativeErrors: field.ErrorList{
+				field.Invalid(replicasPath, nil, "regular error A"),
+				createCoveredError(minReadySecondsPath, -1, "covered error B").WithOrigin("minimum"),
+				createCoveredError(replicasPath, nil, "covered error C").WithOrigin("minimum"),
+				field.Invalid(selectorPath, nil, "regular error D"),
+			},
+			declarativeErrors: field.ErrorList{
+				field.Invalid(replicasPath, nil, "covered error C").WithOrigin("minimum"),
+			},
+			expectMismatches: true,
+			expectDetailsContaining: []string{
+				"Unexpected difference between hand written validation and declarative validation error results",
+				"unmatched error(s) found",
+				"minReadySeconds",
+			},
+		},
+		{
+			name: "Non-matching subset of imperative w/ covered and declarative errors- extra declarative error",
+			imperativeErrors: field.ErrorList{
+				field.Invalid(replicasPath, nil, "regular error A"),
+				createCoveredError(minReadySecondsPath, -1, "covered error B").WithOrigin("minimum"),
+				createCoveredError(replicasPath, nil, "covered error C").WithOrigin("minimum"),
+				field.Invalid(selectorPath, nil, "regular error D"),
+			},
+			declarativeErrors: field.ErrorList{
+				field.Invalid(minReadySecondsPath, -1, "covered error B").WithOrigin("minimum"),
+				field.Invalid(replicasPath, nil, "covered error C").WithOrigin("minimum"),
+				field.Invalid(containersPath, nil, "extra declarative error E").WithOrigin("foo"),
+			},
+			expectMismatches: true,
+			expectDetailsContaining: []string{
+				"Unexpected difference between hand written validation and declarative validation error results",
+				"extra error(s) found",
+				"containers",
+			},
+		},
```

Comment:
Can we test the edges?  I'm confident when I look at both the tests and implementation code that this is right, but when I look at just the tests, I'm less confident, and I want to not only be sure this code is right, but also make sure future edits don't break it.  How about we cover:

- declarative and imperative return 0 errors
- declarative returns multiple errors with origins and the origins are not all the same (match and non-match case)
- declarative and imperative return exactly 1 error (match and non-match cases)
- declarative returns 0 errors, imperative returns 1 error (match and non-match cases)
- declarative returns 1 errors, imperative returns 0 error (match and non-match cases)

Reply:
Defining some fields before declaring the test grid, like `errA = field.Invalid(minReadySecondsPath, -1, "covered error B").WithOrigin("minimum")` might make the tests easier to read/write too

Reply:
I have refactored so that the errs are variables and added test cases for the suggested edge cases mentioned